smolyay/surrogate.py

```diff
@@ -0,0 +1,309 @@
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points to be sampled."""
+        if self._points is None:
+            self._points = self._mapdomain(
+                self.grid.points, [[-1, 1]]*self.dimension, self._domain)
+        return self._points
+
+    def __call__(self, x):
+        """Evaluate surrogate at a given input.
+
+        Parameter
+        ---------
+        x: list (for dimension > 1), or float (for dimension = 1)
+            Input.
+
+        Returns
+        -------
+        float
+            Surrogate output at x.
+
+        Raises
+        ------
+        ValueError
+            For surrogate to be evaluated, function needs to be trained.
+        IndexError
+            Input of the surrogate must be of length of the
+            function's dimensionality.
+        """
+        if self._coefficients is None:
+            raise ValueError('Function needs training!')
+        x = numpy.array(x, copy=False, ndmin=1)
+        if x.shape != (self.dimension, ):
+            raise IndexError('Input must match dimension of domain.')
+
+        # transform point into basis domain and evaluate
+        x_scaled = self._mapdomain(x, self._domain, [[-1, 1]]*self.dimension)
+        value = 0
+        for coeff, basis in zip(self._coefficients, self.grid.basis_functions):
+            if self.dimension > 1:
+                term = numpy.product(
+                    [basis_i(x_i) for basis_i, x_i in zip(basis, x_scaled)])
+            else:
+                term = basis(x_scaled)
+            value += coeff*term
+
+        return value
+
+    def train(self, function, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to the function.
+
+        Parameters
+        ----------
+        function: callabe
+            Function to be approximated.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition.
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+        """
+        data = [function(point) for point in self.points]
+        self.train_from_data(data, linear_solver)
+
+    def train_from_data(self, data, linear_solver='lu'):
+        """Fit surrogate's components (basis functions) to data.
+
+        Parameters
+        ----------
+        data: list
+            function at grid points.
+        linear_slover: string
+            Method of solving linear equations:
+                'lu': lower upper decomposition
+                'inv': solve for coefficients through x = A^{-1}B.
+                'lstsq': least-square solution.
+
+        Raises
+        ------
+        IndexError
+            Data should be of length of the grid points.
+        ValueError
+            Solver should be selected between defined methods.
+        """
+        data = numpy.array(data, copy=False, ndmin=1)
+        if data.shape != (self.points.shape[0],):
```

Comment:
Using `grid.points` here will allow us to bypass creating `points` if it is not actually needed:
```suggestion
        if data.shape != (self.grid.points.shape[0],):
```